doc/BMIconventions.md

```diff
@@ -0,0 +1,313 @@
+
+## Calling `get_value` on input variables
+
+The ngen process *may* call `get_value` on a variable that is returned from `get_input_var_names`. A model *should* return the last value provided to the input variable, unmodified, in this case. At a minimum, a model *must* not crash if this occurs, and should return some reasonable value, rather than throwing an exception or returning BMI_FAILURE (throwing an exception or returning BMI_FAILURE *is* however an appropriate and preferred response if no value has yet been set for the variable).
+
+## Input and output with the same name not permitted
+
+A single variable name cannot be published in both `get_var_input_names` and `get_var_output_names`--this will cause a validation error. Furthermore, it is not presently possible to work around this situation with variable name mapping, because <!-- TODO: This may change with a config file format refactor? --> a variable name mapping does not specify whether it is an input or output variable and could therefore not map only one or the other.
+
+# Unpublished variables
+
+As discussed above, it is not defined by the BMI specification that you may *only* call `get_value` and `set_value` on variable names returned by either `get_input_var_names` or `get_output_var_names`. We refer to variables that appear in neither list--but are understood by a BMI module--as "unpublished" variables. Because ngen uses the input and output variable listings to validate formulations, there are many use cases for such variables, for instance:
+
+* Initialization parameters
+* Calibratable parameters
+* State inspection (for serialization or debugging)
+* Metadata variables (spatial reference information, hyperparameters, etc.) 
+
+These and other uses may be applied by ngen. Specifically, ngen will make use of specified, reserved unpublished variables enumerated later in this document. Therefore, a module should be able to communicate via unpublished variables or at least not malfunction in the case of unpublished variables being used, as described below.
+
+Note that there is currently no discoverability mechanism for unpublished variables, they must be described in documentation (such as this document or a module's  documentation) and often manually implemented or configured for use (such as in a realization config file).
+
+## Unknown/unexpected variables passed to `set_value`
+
+Importantly, if an unexpected variable name is passed to `set_value` this *must* not cause an error or instability! Consider the unknown unpublished value to be "offered" and it can be ignored<!-- TODO: " (unless otherwise specified for specific variable names later in this document)"? We might have a required one at some point. -->. Generally, the preferred behavior in the case of an unknown variable is to do nothing and report success (i.e. throw no exception and return `BMI_SUCCESS`).
```

Comment:
```suggestion
Importantly, if an unexpected variable name is passed to `set_value` this *must* not cause an error or instability! Consider the unknown unpublished value to be "offered" and it can be ignored<!-- TODO: " (unless otherwise specified for specific variable names later in this document)"? We might have a required one at some point. -->. Generally, the preferred behavior in the case of an unknown variable is to do nothing (i.e. throw no exception) and return `BMI_FAILURE`&mdash;ngen will interpret this return value as indicating that the variable just set is not supported by the module.
```